src/ripple/consensus/Consensus.h

```diff
@@ -0,0 +1,1746 @@
+    for (auto& pit : peerProposals_)
+    {
+        auto cit (acquired_.find (pit.second.position ()));
+
+        if (cit != acquired_.end ())
+            dtx.setVote (pit.first,
+                cit->second.exists (txID));
+    }
+
+    impl().relay(dtx);
+
+    disputes_.emplace (txID, std::move (dtx));
+}
+
+template <class Derived, class Traits>
+void Consensus<Derived, Traits>::adjustCount (TxSet_t const& txSet,
+    std::vector<NodeID_t> const& peers)
+{
+    for (auto& it : disputes_)
+    {
+        bool setHas = txSet.exists (it.first);
+        for (auto const& pit : peers)
+            it.second.setVote (pit, setHas);
+    }
+}
+
+/** How many of the participants must agree to reach a given threshold?
+
+    Note that the number may not precisely yield the requested percentage.
+    For example, with with size = 5 and percent = 70, we return 3, but
+    3 out of 5 works out to 60%. There are no security implications to
+    this.
+
+    @param participants The number of participants (i.e. validators)
+    @param percent The percent that we want to reach
+
+    @return the number of participants which must agree
+*/
+inline int
+participantsNeeded (int participants, int percent)
+{
+    int result = ((participants * percent) + (percent / 2)) / 100;
+
+    return (result == 0) ? 1 : result;
+}
+
+template <class Derived, class Traits>
+void Consensus<Derived, Traits>::updateOurPositions ()
+{
+    // Compute a cutoff time
+    auto const peerCutoff = now_ - PROPOSE_FRESHNESS;
+    auto const ourCutoff = now_ - PROPOSE_INTERVAL;
+
+    // Verify freshness of peer positions and compute close times
+    std::map<NetTime_t, int> closeTimes;
+    {
+        auto it = peerProposals_.begin ();
+        while (it != peerProposals_.end ())
+        {
+            if (it->second.isStale (peerCutoff))
+            {
+                // peer's proposal is stale, so remove it
+                auto const& peerID = it->second.nodeID ();
+                JLOG (j_.warn())
+                    << "Removing stale proposal from " << peerID;
+                for (auto& dt : disputes_)
+                    dt.second.unVote (peerID);
+                it = peerProposals_.erase (it);
+            }
+            else
+            {
+                // proposal is still fresh
+                ++closeTimes[effectiveCloseTime(it->second.closeTime(),
+                    closeResolution_, previousLedger_.closeTime())];
+                ++it;
+            }
+        }
+    }
+
+    // This will stay unseated unless there are any changes
+    boost::optional <TxSet_t> ourNewSet;
+
+    // Update votes on disputed transactions
+    {
+        for (auto& it : disputes_)
+        {
+            // Because the threshold for inclusion increases,
+            //  time can change our position on a dispute
+            if (it.second.updateVote (closePercent_, proposing_))
+            {
+                if (!ourNewSet)
+                    ourNewSet = *ourSet_;
+
```

Comment:
Somehow, this isn't taking a mutable snapshot, so when we call `insert` or `erase`, we're trying to modify an immutable, shared map.